### PR TITLE
ci: test/end2end_tests

### LINKED ISSUE #166: SSHNP Automated e2e tests
### Is your feature request related to a problem? Please describe.

We do not currently have any automated end to end tests, so all testing is manually intensive and time-consuming

### Describe the solution you'd like

A set of tests of the following scenarios
- With 'new' atSigns, full scenario, single sshnp client - establish ssh session via sshrv, verify all worked and cleaned up correctly
- With 'existing' atSigns, same scenario
- For both of the above scenarios, with multiple concurrent sshnp clients
- Compatibility testing for all of the above - i.e. sshnpd from previous releases vs sshnp from latest trunk, sshnpd from latest trunk vs sshnp from previous releases, etc

### Describe alternatives you've considered

_No response_

### Additional context

Things that might help:
- There are tests in at_server which create, activate and onboard new atSigns in the staging environment, run the at_server e2e test pack using those atSigns, and then delete those atSigns at the end
- There has been work done on dockerizing sshnpd, sshnp and sshrv

### PR DESCRIPTION
**Pull request by JeremyTubongbanua**
## What I did

- Automated end-to-end testing `.github/workflows/end2end_tests.yaml`

Additionally, a few tools for testing ssh no ports locally:

- `tools/automated-tests-locally` - spin up containers that will run entrypoint.sh on their own
- `tools/manual-tool` - spin up containers interactively with binaries in ~/.local/bin

closes #166

### CODE REVIEW

**Comment by JeremyTubongbanua** [comment]
Proof of it working

image

File: packages/sshnoports/test/end2end_tests/contexts/sshnp/keys/placeholder [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This is to ensure the keys/ directory exists.
```
Comment (xavierchanth) [inline packages/sshnoports/test/end2end_tests/contexts/sshnp/keys/placeholder:1]: Use a `.gitkeep` file for best practices

File: packages/sshnoports/test/end2end_tests/contexts/sshnpd/keys/placeholder [hunk 2]
```diff
@@ -0,0 +1,3 @@
+This is to ensure the keys/ directory exists.
```
Comment (xavierchanth) [inline packages/sshnoports/test/end2end_tests/contexts/sshnpd/keys/placeholder:1]: Same thing here

File: packages/sshnoports/test/end2end_tests/contexts/sshrvd/keys/placeholder [hunk 3]
```diff
@@ -0,0 +1,3 @@
+This is to ensure the keys/ directory exists.
```
Comment (xavierchanth) [inline packages/sshnoports/test/end2end_tests/contexts/sshrvd/keys/placeholder:1]: And here

**Review by xavierchanth — APPROVED** [review]
Other than the comments below LGTM

**Review by gkc — APPROVED** [review]
Great work @JeremyTubongbanua - will approve as it stands and you can follow up with more PRs as you see fit

**Pull request merged**